src/Crypto/Encoding/BIP39/BIP39Err.hs

```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+module Crypto.Encoding.BIP39.BIP39Err (BIP39Err(..)) where
+
+import           Foundation hiding (show)
+import           Prelude    (Show, show)
+
+data BIP39Err
+    = BIP39ErrInvalidDictionaryWord String
+
+    | BIP39ErrInvalidEntropyLength
+        Int -- ^ Actual length
+        Int -- ^ Expected length
+
+    | BIP39ErrInvalidEntropyChecksum
+        Word8 -- ^ Actual checksum
+        Word8 -- ^ Expected checksum
+
+instance Show BIP39Err where
+    show = \case
+        BIP39ErrInvalidDictionaryWord w ->
+          "Invalid word: " <> show w <> ", not found in dictionary"
+
+        BIP39ErrInvalidEntropyLength a e ->
+          "Invalid entropy length: got " <> show a <>
+          " words, but expected " <> show e
+
+        BIP39ErrInvalidEntropyChecksum a e ->
+          "Invalid entropy checksum: got " <> show a <>
+          " but expected " <> show e
+
```

Comment:
Now that I look at the code, I am wondering whether we couldn't have:

```hs
data EntropyErr = ErrInvalidEntropyLength | ErrInvalidEntropyChecksum

-- and

data DictionaryErr = ErrInvalidDictionaryWord
```

We lose a bit of the type safety by having them grouped in one type. There's no reason why `wordtoIndex` would return an error that could be `ErrInvalidEntropyLength`; And most (all?) functions actually returning `ErrInvalidEntropyLength` or `ErrInvalidEntropyCheckum` have nothing to do with the dictionary either.

Note that I've also dropped the `BIP39` prefix for I think they adds a bit of unnecessary overhead here.